agent/internal/fluent_intg_test.go

```diff
@@ -0,0 +1,258 @@
+		body, err := ioutil.ReadAll(ctx.Request().Body)
+		if err != nil {
+			return err
+		}
+
+		var logs []model.TrialLog
+		if err = json.Unmarshal(body, &logs); err != nil {
+			return err
+		}
+
+		for _, l := range logs {
+			logBuffer <- l
+		}
+		return nil
+	})
+	go func() {
+		if err := e.StartServer(e.Server); err != nil {
+			t.Logf("mock master exited: %s", err)
+		}
+	}()
+	return logBuffer, func() {
+		if err := e.Close(); err != nil {
+			t.Logf("error closing mock master echo server: %s", err)
+		}
+		if err := e.Listener.Close(); err != nil {
+			t.Logf("error closing mock master listener: %s", err)
+		}
+	}
+}
+
+func runContainerWithLogs(t *testing.T, fakeLogs string, trialID, fluentPort int) {
+	// WHEN we start a container that logs with fluentbit as its log driver.
+	docker, err := client.NewClientWithOpts(client.FromEnv, client.WithVersion("1.40"))
+	assert.NilError(t, err, "error connecting to Docker daemon")
+
+	logs, err := docker.ImagePull(context.Background(), "busybox", types.ImagePullOptions{})
+	assert.NilError(t, err, "failed to pull image")
+	_, err = ioutil.ReadAll(logs)
+	assert.NilError(t, err, "failed to drain pull logs")
+
+	// AND set some known ENV values.
+	spec := cproto.Spec{
+		RunSpec: cproto.RunSpec{
+			ContainerConfig: container.Config{
+				Cmd:   []string{"cat", "/fakelogs"},
+				Image: "busybox",
+			},
+			UseFluentLogging: true,
+		},
+	}
+	env := []string{fmt.Sprintf("DET_TRIAL_ID=%d", trialID)}
+	cont := cproto.Container{ID: "goodcontainer"}
+	spec = overwriteSpec(cont, spec, env, nil, fluentPort)
+
+	// AND create a container with this env
+	cc, err := docker.ContainerCreate(
+		context.Background(),
+		&spec.RunSpec.ContainerConfig,
+		&spec.RunSpec.HostConfig,
+		&spec.RunSpec.NetworkingConfig,
+		"log-test-"+uuid.New().String())
+	assert.NilError(t, err, "error creating container")
+
+	// AND start that container with a command to print the fake logs
+	files := []archive.Item{
+		{
+			Path:     "/fakelogs",
+			Type:     tar.TypeReg,
+			FileMode: 0444,
+			Content:  []byte(fakeLogs),
+		},
+	}
+	filesReader, err := archive.ToIOReader(files)
+	assert.NilError(t, err, "failed make reader from fluent files")
+	err = docker.CopyToContainer(context.Background(),
+		cc.ID,
+		"/",
+		filesReader,
+		types.CopyToContainerOptions{},
+	)
+	assert.NilError(t, err, "failed to copy files to container")
+	err = docker.ContainerStart(context.Background(), cc.ID, types.ContainerStartOptions{})
+	assert.NilError(t, err, "error starting container")
+}
+
+func assertLogEquals(t *testing.T, l, expected model.TrialLog) {
+	assert.Equal(t, l.TrialID, expected.TrialID, spew.Sdump(l))
+	assert.DeepEqual(t, l.AgentID, expected.AgentID)
+	assert.DeepEqual(t, l.ContainerID, expected.ContainerID)
+	assert.DeepEqual(t, l.RankID, expected.RankID)
+	assert.DeepEqual(t, l.Level, expected.Level)
+	assert.DeepEqual(t, l.Log, expected.Log)
+	assert.DeepEqual(t, l.StdType, expected.StdType)
```

Comment:
blocking: `DeepEqual` will display detailed differences for struct fields on failure, so I think this function can be replaced by a single call to it. (Sorry, I may have sort of suggested the one-per-field thing before, but it shouldn't be necessary.)

```
          model.TrialLog{
                ID:          0,
        -       TrialID:     17,
        +       TrialID:     18,
                Message:     "",
                AgentID:     nil,
                ContainerID: nil,
                RankID:      nil,
                Timestamp:   nil,
        -       Level:       &"INFO",
        +       Level:       nil,
                Log:         nil,
                Source:      nil,
        -       StdType:     nil,
        +       StdType:     &"stdout",
          }
```

Reply:
Changed to nil'ing out `Timestamp` and using `DeepEqual` on the whole object on the grounds that if we add a field to the trial logs, we'll see this test break and someone will have to add tests for their work 😄